### PR TITLE
Refactor \Order\Shipment\Save Controller to use ResultInterface

### PR DESCRIPTION
**Pull request by Jeroenll**
### Description (*)
Refactor \Order\Shipment\Save Controller to use ResultInterface

### Fixed Issues (if relevant)
1. Found references to internal ticket MAGETWO-28356: Refactor controller actions to new ResultInterface.

### Manual testing scenarios (*)
1. Create shipment and see if redirects are working.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @JeroenVanLeusden. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by ihor-sviziev** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, here is your [new Magento instance](https://pr-20057.instances.magento-community.engineering).
Admin access: [https://pr-20057.instances.magento-community.engineering/admin](https://pr-20057.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php [hunk 1]
```diff
@@ -169,38 +181,23 @@ public function execute()
...
                 $responseAjax->setMessage(__('An error occurred while creating shipping label.'));
             } else {
-                $this->messageManager->addError(__('Cannot save shipment.'));
-                $this->_redirect('*/*/new', ['order_id' => $this->getRequest()->getParam('order_id')]);
+                $this->messageManager->addErrorMessage(__('Cannot save shipment.'));
+                return $resultRedirect->setPath('*/*/new', ['order_id' => $this->getRequest()->getParam('order_id')]);
             }
         }
         if ($isNeedCreateLabel) {
             $this->getResponse()->representJson($responseAjax->toJson());
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: Looks like there will be regression. Earlier redirect wasn't sent.
@aredridel  could you re-check it?

Also there looks like should be returned result object with json.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php [hunk 2]
```diff
@@ -118,8 +131,9 @@ public function execute()
             $this->shipmentLoader->setTracking($this->getRequest()->getParam('tracking'));
             $shipment = $this->shipmentLoader->load();
             if (!$shipment) {
-                $this->_forward('noroute');
-                return;
+                $resultForward = $this->resultForwardFactory->create();
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: Looks like we can get resultForward without getting one more dependency:

Here is example:
https://github.com/magento/magento2/blob/df2784b6dec6fe841c72223705974231c246a465/app/code/Magento/Catalog/Controller/Adminhtml/Product/Reload.php#L22

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @paliarush,
I'd like to mark https://github.com/magento/magento2/blob/b475068f894b95674454cc08dc327a45d218a72f/app/code/Magento/Backend/App/AbstractAction.php#L323-L337 and https://github.com/magento/magento2/blob/b475068f894b95674454cc08dc327a45d218a72f/app/code/Magento/Backend/App/AbstractAction.php#L308-L321

as deprecated with description that we need to use result object instead. What do you think? Can we add deprecation for them?

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php (hunk 2 above)
Comment (Jeroenll) [reply app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: @ihor-sviziev Updated the PR with this change.

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php [hunk 3]
```diff
@@ -43,22 +43,36 @@ class Save extends \Magento\Backend\App\Action implements HttpPostActionInterfac
...
     ) {
+        parent::__construct($context);
+
         $this->shipmentLoader = $shipmentLoader;
         $this->labelGenerator = $labelGenerator;
         $this->shipmentSender = $shipmentSender;
-        parent::__construct($context);
+        $this->shipmentValidator = $shipmentValidator ?: \Magento\Framework\App\ObjectManager::getInstance()
+            ->get(\Magento\Sales\Model\Order\Shipment\ShipmentValidatorInterface::class);
+        $this->resultForwardFactory = $resultForwardFactory ?: \Magento\Framework\App\ObjectManager::getInstance()
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: You added this dependency, but not using it anywhere.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Looks like here still few issues:
1. added not needed dependency: https://github.com/magento/magento2/pull/20057#discussion-diff-246268446R74
2. here is regression, previously there were if-else, but now it will execute both parts.
https://github.com/magento/magento2/pull/20057#discussion-diff-245016993R197
3. In one place there still NOT return result object, but setting response body via ```$this->getResponse()->representJson($responseAjax->toJson()).```. Would be better to fix all places to return ResultInterface
https://github.com/magento/magento2/pull/20057#discussion-diff-245016993R198
4. There are some unit tests failures. Please fix them.

There also integration tests failures, but I think they should NOT be related to this change, so it's actually ok.

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php (hunk 1 above)
Comment (aredridel) [reply app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: I'm not sure I follow. Re-check what?

**Comment by Jeroenll** [comment]
@ihor-sviziev Should be all solved now.

File: app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php [hunk 4]
```diff
@@ -86,7 +91,7 @@ protected function _saveShipment($shipment)
      * Save shipment
      * We can save only new shipment. Existing shipments are not editable
      *
-     * @return void
+     * @return \Magento\Backend\Model\View\Result\Redirect
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/Save.php]: In one case there will be result Jason, so I prefer to write there result interface. It’s really minor, I can fix it myself

**Review by ihor-sviziev — COMMENTED** [review]
Now everything is fine, will fix one really minor issue little bit later and approve your pr

**Review by ihor-sviziev — APPROVED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-3831](https://jira.corp.magento.com/browse/ENGCOM-3831) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
Hi @JeroenVanLeusden. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.

**Pull request merged**